tensorly/decomposition/candecomp_parafac.py

```diff
@@ -1,66 +1,152 @@
+    ----------
+    factors : ndarray list
+        list of matrices, all with the same number of columns
+        i.e.::
+            for u in U:
+                u[i].shape == (s_i, R)
+
+        where `R` is fixed while `s_i` can vary with `i`
+
+    Returns
+    -------
+    normalized_factors : list of ndarrays
+        list of matrices with the same shape as `factors`
+    weights : ndarray
+        vector of length `R` holding normalizing constants
+
+    """
+    # allocate variables for weights, and normalized factors
+    rank = factors[0].shape[1]
+    weights = T.ones(rank)
+    normalized_factors = []
+
+    # normalize columns of factor matrices
+    for factor in factors:
+        scales = T.norm(factor, axis=0)
+        weights *= scales
+        scales_non_zero = T.where(scales==0, T.ones(T.shape(scales)), scales)
+        normalized_factors.append(factor/scales_non_zero)
+    return normalized_factors, weights
+
+
+def initialize_factors(tensor, rank, init='svd', random_state=None):
+    r"""Initialize factors used in `parafac`.
 
-        Computes a rank-`rank` decomposition of `tensor` [1]_ such that:
-        ``tensor = [| factors[0], ..., factors[-1] |]``
+    The type of initialization is set using `init`. If `init == 'random'` then
+    initialize factor matrices using `random_state`. If `init == 'svd'` then
+    initialize the `m`th factor matrix using the `rank` left singular vectors
+    of the `m`th unfolding of the input tensor.
 
     Parameters
     ----------
     tensor : ndarray
-    rank  : int
-            number of components
-    n_iter_max : int
-                 maximum number of iteration
+    rank : int
     init : {'svd', 'random'}, optional
-    tol : float, optional
-          tolerance: the algorithm stops when the variation in
-          the reconstruction error is less than the tolerance
-    random_state : {None, int, np.random.RandomState}
-    verbose : int, optional
-        level of verbosity
 
     Returns
     -------
     factors : ndarray list
-            list of factors of the CP decomposition
-            element `i` is of shape (tensor.shape[i], rank)
+        List of initialized factors of the CP decomposition where element `i`
+        is of shape (tensor.shape[i], rank)
 
-    References
-    ----------
-    .. [1] T.G.Kolda and B.W.Bader, "Tensor Decompositions and Applications",
-       SIAM REVIEW, vol. 51, n. 3, pp. 455-500, 2009.
     """
     rng = check_random_state(random_state)
 
     if init is 'random':
-        factors = [T.tensor(rng.random_sample((tensor.shape[i], rank)), **T.context(tensor)) for i in range(T.ndim(tensor))]
+        logger.info('[CPDecomp] Using random initialization')
+        factors = [T.tensor(rng.random_sample((tensor.shape[i], rank))) for i in range(T.ndim(tensor))]
+        factors, _ = normalize_factors(factors)
+        return factors
 
-    elif init is 'svd':
+    if init is 'svd':
+        logger.info('[CPDecomp] Using svd initialization')
```

Comment:
i would just use a print, using the `verbose` parameter that should be propagated from the main function

Reply:
As you wish. However, I recommend looking into the `logging` module for controlling what sort of statements are sent to stdout. The opposing argument to using `verbose` parameter is that it adds clutter to the argument list (every related function and sub-function would need to handle verbosity) and there are no built-in level controls: during "normal" operation you may want to only print high-level information but while debugging you may want to run the script with lower level output.

But since no other part of Tensorly uses logging I can remove it.